### PR TITLE
Standalone Lite: Accelerators

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds the Accelerator base class and CPU, TPU, CUDA, MPS implementations.

- The base class has minor changes: 1) setup(Trainer) removed 2) setup_environment renamed to init_device
- PL's Accelerator base class now subclasses Lite's Accelerator class. 
- Later on, PL can refactor their usage in strategies to completely rely on the implementations in Lite. The small blocker here currently is the usage of the trainer reference. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: tests/tests_pytorch/accelerators/test_registry.py [hunk 1]
```diff
@@ -11,55 +11,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from pytorch_lightning import Trainer
-from pytorch_lightning.accelerators import Accelerator, AcceleratorRegistry
-
-
-def test_accelerator_registry_with_new_accelerator():
```
Comment (awaelchli) [inline tests/tests_pytorch/accelerators/test_registry.py:18]: Moved this test. It does not need to be duplicated because what is tested here is independent of where the registry is used

File: tests/tests_pytorch/accelerators/test_registry.py [hunk 2]
```diff
@@ -11,55 +11,7 @@
...
-    trainer = Trainer(accelerator=accelerator_name, devices="auto")
-    assert isinstance(trainer.accelerator, CustomAccelerator)
-    assert trainer.strategy.parallel_devices == ["foo"] * 3
-
-    AcceleratorRegistry.remove(accelerator_name)
-    assert accelerator_name not in AcceleratorRegistry
+from pytorch_lightning.accelerators import AcceleratorRegistry
 
 
 def test_available_accelerators_in_registry():
```
Comment (awaelchli) [inline tests/tests_pytorch/accelerators/test_registry.py:17]: This is kept because Lite and PL support a different set of accelerators (hence separate registries)

File: tests/tests_lite/utilities/test_device_parser.py [hunk 3]
```diff
@@ -27,5 +28,12 @@
...
-    assert lightning_lite.utilities.device_parser.num_cuda_devices() == 2
+    assert device_parser.is_cuda_available()
+    assert device_parser.num_cuda_devices() == 2
+
+
+@pytest.mark.parametrize("devices", ([3], -1))
+def test_invalid_devices_with_cpu_accelerator(devices):
+    """Test invalid device flag raises MisconfigurationException."""
+    with pytest.raises(MisconfigurationException, match="should be an int > 0"):
+        device_parser.parse_cpu_cores(devices)
```
Comment (awaelchli) [inline tests/tests_lite/utilities/test_device_parser.py:39]: moved here from tests/pytorch/accelerators/test_cpu.py

File: tests/tests_lite/accelerators/test_tpu.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    [
+        (0, []),  # TODO(lite): This should raise an exception
+        (1, [0]),
+        (2, [0, 1]),
+        (3, [0, 1, 2]),
+        ("anything-else", "anything-else"),  # TODO(lite): This should raise an exception
+    ],
+)
+def test_get_parallel_devices(devices, expected):
+    assert TPUAccelerator.get_parallel_devices(devices) == expected
```
Comment (awaelchli) [inline tests/tests_lite/accelerators/test_tpu.py:41]: Awkward cases that were not handled in PL

File: tests/tests_lite/accelerators/test_cuda.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
+# Copyright The PyTorch Lightning team.
```
Comment (awaelchli) [inline tests/tests_lite/accelerators/test_cuda.py:1]: A mix of test cases carried over from tests/pytorch/accelerators/test_gpu.py and newly written ones

File: tests/tests_lite/accelerators/test_cpu.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+# Copyright The PyTorch Lightning team.
```
Comment (awaelchli) [inline tests/tests_lite/accelerators/test_cpu.py:1]: A mix of test cases carried over from tests/pytorch/accelerators/test_cpu.py and newly written ones

File: src/pytorch_lightning/accelerators/accelerator.py [hunk 7]
```diff
@@ -31,54 +29,9 @@ class Accelerator(ABC):
...
-        This is called before the LightningModule/DataModule setup hook which allows the user to access the accelerator
-        environment before setup is complete.
-        """
-
     def setup(self, trainer: "pl.Trainer") -> None:
         """Setup plugins for the trainer fit and creates optimizers.
 
         Args:
             trainer: the trainer instance
         """
```
Comment (awaelchli) [inline src/pytorch_lightning/accelerators/accelerator.py:37]: This trainer-specific method is currently the only difference between the Accelerator base class in PL vs Lite. In the future, we will be able to refactor the usages to merge the base classes.

File: src/pytorch_lightning/accelerators/accelerator.py [hunk 8]
```diff
@@ -11,18 +11,16 @@
...
-import torch
+from abc import ABC
 
 import pytorch_lightning as pl
+from lightning_lite.accelerators.accelerator import Accelerator as _Accelerator
 
 
-class Accelerator(ABC):
-    """The Accelerator Base Class. An Accelerator is meant to deal with one type of Hardware.
+class Accelerator(_Accelerator, ABC):
```
Comment (awaelchli) [inline src/pytorch_lightning/accelerators/accelerator.py:20]: Note: Inherits from Lite Accelerator base class

File: src/lightning_lite/accelerators/registry.py [hunk 9]
```diff
@@ -112,11 +112,8 @@ def __str__(self) -> str:
         return "Registered Accelerators: {}".format(", ".join(self.available_accelerators()))
 
 
-AcceleratorRegistry = _AcceleratorRegistry()
-
-
-def call_register_accelerators(base_module: str) -> None:
```
Comment (awaelchli) [inline src/lightning_lite/accelerators/registry.py:118]: This now takes the registry as input, because there are two now: One for Lite, one for PL.

File: src/lightning_lite/accelerators/accelerator.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+    - CPU
+    - CUDA (NVIDIA)
+    - MPS (Apple Silicon)
+    - TPU (Google)
+    """
+
+    @abstractmethod
+    def init_device(self, device: torch.device) -> None:
+        """Create and prepare the device for the current process."""
```
Comment (awaelchli) [inline src/lightning_lite/accelerators/accelerator.py:33]: Previously, this was named setup_environment. Renamed to `init_device` to make the intent of this method clear and focused. 

The rest of the interface was adapted 1:1 from PL. All methods are now marked abstract.

File: src/lightning_lite/accelerators/accelerator.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+class Accelerator(ABC):
+    """The Accelerator base class. An Accelerator is meant to deal with one type of hardware.
+
+    Currently, there are accelerators for:
+
+    - CPU
+    - CUDA (NVIDIA)
+    - MPS (Apple Silicon)
+    - TPU (Google)
```
Comment (carmocca) [inline src/lightning_lite/accelerators/accelerator.py:28]: I guess this was copied over, but I suggest we don't list them in a docstring.
This can easily be forgotten, and other accelerators can be registered through the registry

```suggestion
```

File: src/lightning_lite/accelerators/cpu.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        """Get CPU stats from ``psutil`` package."""
+        return get_cpu_stats()
+
+    def teardown(self) -> None:
+        pass
+
+    @staticmethod
+    def parse_devices(devices: Union[int, str, List[int]]) -> int:
+        """Accelerator device parsing logic."""
+        devices = device_parser.parse_cpu_cores(devices)
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cpu.py:45]: (maybe not in this PR): I would move these utility functions out of `device_parser` into the accelerator that uses them.

I understand we keep them because the PL accelerator will want to use them too, but PL can import them from this Lite module. Code locality is good! Say no to random utility files!

File: src/pytorch_lightning/accelerators/accelerator.py [hunk 13]
```diff
@@ -11,18 +11,19 @@
...
+    """The Accelerator base class for Lightning PyTorch. An Accelerator is meant to deal with one type of hardware.
 
-    Currently there are accelerators for:
+    Currently, there are accelerators for:
 
     - CPU
```
Comment (carmocca) [inline src/pytorch_lightning/accelerators/accelerator.py:28]: Same suggestion about removing this list

File: src/pytorch_lightning/accelerators/cpu.py [hunk 14]
```diff
@@ -70,22 +73,9 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
...
-    return {
-        _CPU_VM_PERCENT: psutil.virtual_memory().percent,
-        _CPU_PERCENT: psutil.cpu_percent(),
-        _CPU_SWAP_PERCENT: psutil.swap_memory().percent,
-    }
+    rank_zero_deprecation(
+        "`pytorch_lightning.accelerators.cpu.get_cpu_stats` has been deprecated in v1.8.0 and will be removed in"
+        " v1.10.0. Please use `lightning_lite.accelerators.cpu.get_cpu_stats` instead."
+    )
+    return new_get_cpu_stats()
```
Comment (carmocca) [inline src/pytorch_lightning/accelerators/cpu.py:81]: I think all the stats logic should stay in PL, removed from the Lite implementations.

This is for the DeviceStats callback, and Lite won't have callbacks

File: src/lightning_lite/accelerators/accelerator.py (hunk 10 above)
Comment (awaelchli) [reply src/lightning_lite/accelerators/accelerator.py:33]: > Does that mean you will rename setup_environment for the PL accelerators?

Yes. We need to make the hooks and APIs in Lite very focused and unambiguous. It is guided by the flexibility we need to maintain in Lite, as there are less assumptions than in PL. This also makes the code more readable. setup_environment was named because of where and how it was called in Lite, prior to accelerator being under strategy.

File: src/lightning_lite/accelerators/cpu.py (hunk 12 above)
Comment (awaelchli) [reply src/lightning_lite/accelerators/cpu.py:45]: Yes

File: src/lightning_lite/accelerators/mps.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+from lightning_lite.accelerators.accelerator import Accelerator
+from lightning_lite.utilities import device_parser
+from lightning_lite.utilities.imports import _PSUTIL_AVAILABLE, _TORCH_GREATER_EQUAL_1_12
+
+# For using the `MPSAccelerator`, user's machine should have `torch>=1.12`, Metal programming framework and
+# the ARM-based Apple Silicon processors.
+_MPS_AVAILABLE = (
+    _TORCH_GREATER_EQUAL_1_12 and torch.backends.mps.is_available() and platform.processor() in ("arm", "arm64")
+)
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning_lite/accelerators/mps.py:27]: We need to define it somewhere so we can import it in other places. In https://github.com/Lightning-AI/lightning/issues/12786 I suggested to make functions, but we didn't go this route unfortunately.

[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning_lite/accelerators/mps.py:27]: `MPSAccelerator.is_available` could be imported elsewhere

File: src/lightning_lite/accelerators/cuda.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    @staticmethod
+    def is_available() -> bool:
+        return device_parser.num_cuda_devices() > 0
+
+    @classmethod
+    def register_accelerators(cls, accelerator_registry: Dict) -> None:
+        accelerator_registry.register(
+            "cuda",
+            cls,
+            description=f"{cls.__class__.__name__}",
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:63]:
```suggestion
            description=cls.__class__.__name__",
```

File: src/pytorch_lightning/accelerators/cuda.py [hunk 17]
```diff
@@ -98,12 +100,8 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
             description=f"{cls.__class__.__name__}",
         )
 
-    def teardown(self) -> None:
-        # clean up memory
-        torch.cuda.empty_cache()
-
 
-def get_nvidia_gpu_stats(device: _DEVICE) -> Dict[str, float]:  # pragma: no-cover
```
Comment (carmocca) [inline src/pytorch_lightning/accelerators/cuda.py:106]: Why are we changing these annotations?

**Review by carmocca — APPROVED** [review]
LGTM. Previous comments about moving implementations can be `TODO(lite)`s

**Review by Borda — APPROVED** [review]

File: src/lightning_lite/accelerators/cpu.py [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+class CPUAccelerator(Accelerator):
+    """Accelerator for CPU devices."""
+
+    def init_device(self, device: torch.device) -> None:
+        """
+        Raises:
+            ValueError:
+                If the selected device is not CPU.
+        """
+        if device.type != "cpu":
```
Comment (tchaton) [inline src/lightning_lite/accelerators/cpu.py:31]: What if the device is a meta device ?

File: src/lightning_lite/accelerators/cpu.py [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+import torch
+
+from lightning_lite.accelerators.accelerator import Accelerator
+from lightning_lite.utilities import device_parser
+
+
+class CPUAccelerator(Accelerator):
+    """Accelerator for CPU devices."""
+
+    def init_device(self, device: torch.device) -> None:
```
Comment (tchaton) [inline src/lightning_lite/accelerators/cpu.py:25]: Looks more like a setup.

**Review by tchaton — APPROVED** [review]
LGTM !

File: src/lightning_lite/accelerators/cpu.py (hunk 19 above)
Comment (awaelchli) [reply src/lightning_lite/accelerators/cpu.py]: Changed to setup_device

File: src/lightning_lite/accelerators/cpu.py (hunk 18 above)
Comment (awaelchli) [reply src/lightning_lite/accelerators/cpu.py:31]: Yes, I only moved the code here because this validation is also in PL. Going to leave it like this for now.

File: src/lightning_lite/accelerators/accelerator.py (hunk 10 above)
Comment (awaelchli) [reply src/lightning_lite/accelerators/accelerator.py]: Changed to setup_device
